src/Reflection/Annotations/AnnotationsMethodsClassReflectionExtension.php

```diff
@@ -76,9 +77,92 @@ private function createMethods(ClassReflection $classReflection): array
+			$parametersStringCandidate = trim(trim($match['Parameters'], '()'));
+			$parameters = $this->createMethodParameters($parametersStringCandidate, $typeMap);
+			$isVariadic = $this->detectMethodVariadic($classReflection, $methodName, $parameters);
+			$methods[$methodName] = new AnnotationMethodReflection($methodName, $classReflection, $returnType, $parameters, $isStatic, $isVariadic);
 		}
 		return $methods;
 	}
 
+	/**
+	 * @param string $parametersStringCandidate
+	 * @param \PHPStan\Type\Type[] $typeMap
+	 * @return AnnotationsMethodParameterReflection[]
+	 */
+	private function createMethodParameters(string $parametersStringCandidate, array $typeMap): array
+	{
+		$parameters = [];
+		if (!$parametersStringCandidate) {
+			return $parameters;
+		}
+
+		foreach (preg_split('#\s*,\s*#', $parametersStringCandidate) as $parameter) {
+			if (preg_match('#(?P<IsNullable>\?)?(?:(?P<Type>' . FileTypeMapper::TYPE_PATTERN . ')\s+)?(?P<IsVariadic>...)?(?P<IsPassedByReference>\&)?\$(?P<Name>[a-zA-Z0-9_]+)(?:\s+=\s+(?P<DefaultValue>.+))?#', $parameter, $parameterMatches)) {
+				$name = $parameterMatches['Name'];
+				$typeString = $parameterMatches['Type'];
+				if ($typeString) {
+					if (!isset($typeMap[$typeString])) {
+						continue;
+					}
+					$type = $typeMap[$typeString];
+					if (!empty($parameterMatches['IsNullable'])) {
+						$type = TypeCombinator::addNull($type);
+					}
+				} else {
+					$type = new MixedType();
+				}
+				$isVariadic = !empty($parameterMatches['IsVariadic']);
+				$isPassedByReference = !empty($parameterMatches['IsPassedByReference']);
+				$isOptional = !empty($parameterMatches['DefaultValue']);
+				$parameters[] = new AnnotationsMethodParameterReflection($name, $type, $isPassedByReference, $isOptional, $isVariadic);
+			}
+		}
+
+		return $parameters;
+	}
+
+	/**
+	 * @param ClassReflection $classReflection
+	 * @param string $methodName
+	 * @param AnnotationsMethodParameterReflection[] $parameters
+	 * @return bool
+	 */
+	private function detectMethodVariadic(ClassReflection $classReflection, string $methodName, array $parameters): bool
+	{
+		$isVariadic = false;
+		$possibleVariadicParameterIndex = count($parameters) - 1;
+		foreach ($parameters as $index => $parameter) {
+			if ($parameter->isVariadic()) {
+				if ($index !== $possibleVariadicParameterIndex) {
+					throw new \PHPStan\ShouldNotHappenException(sprintf(
+						'Variadic parameter can be only the last parameter on method %s::%s and parameter %s.',
+						$classReflection->getName(),
+						$methodName,
+						$parameter->getName()
+					));
+				}
+				if ($parameter->isPassedByReference()) {
+					throw new \PHPStan\ShouldNotHappenException(sprintf(
+						'Variadic parameter %s cannot be passed by reference to method %s::%s.',
+						$parameter->getName(),
+						$classReflection->getName(),
+						$methodName
+					));
+				}
+				if ($parameter->isOptional()) {
+					throw new \PHPStan\ShouldNotHappenException(sprintf(
```

Comment:
ShouldNotHappenException result in internal error reported by PHPStan, which we cannot afford when analyzing someone else's code. I'd prefer to stay silent here. Optionally, we can write some rules to detect invalid annotations.